internal/tiltfile/live_update.go

```diff
@@ -0,0 +1,297 @@
+	var workDir string
+	if err := starlark.UnpackArgs(fn.Name(), args, kwargs, "dir", &workDir); err != nil {
+		return nil, err
+	}
+
+	if !filepath.IsAbs(workDir) {
+		return starlark.None, fmt.Errorf("work_dir path '%s' is not absolute", workDir)
+	}
+
+	ret := liveUpdateWorkDirStep{
+		value:    workDir,
+		position: thread.TopFrame().Position(),
+	}
+	s.unconsumedLiveUpdateSteps = append(s.unconsumedLiveUpdateSteps, ret)
+	return ret, nil
+}
+
+func (s *tiltfileState) liveUpdateSync(thread *starlark.Thread, fn *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var localPath, remotePath string
+	if err := starlark.UnpackArgs(fn.Name(), args, kwargs, "local_path", &localPath, "remote_path", &remotePath); err != nil {
+		return nil, err
+	}
+
+	ret := liveUpdateSyncStep{
+		localPath:  s.absPath(localPath),
+		remotePath: remotePath,
+		position:   thread.TopFrame().Position(),
+	}
+	s.unconsumedLiveUpdateSteps = append(s.unconsumedLiveUpdateSteps, ret)
+	return ret, nil
+}
+
+func (s *tiltfileState) liveUpdateRun(thread *starlark.Thread, fn *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var command string
+	var triggers starlark.Value
+	if err := starlark.UnpackArgs(fn.Name(), args, kwargs, "cmd", &command, "trigger?", &triggers); err != nil {
+		return nil, err
+	}
+
+	triggersSlice := starlarkValueOrSequenceToSlice(triggers)
+	var triggerStrings []string
+	for _, t := range triggersSlice {
+		switch t2 := t.(type) {
+		case starlark.String:
+			triggerStrings = append(triggerStrings, s.absPath(string(t2)))
+		default:
+			return nil, fmt.Errorf("run cmd '%s' triggers contained value '%s' of type '%s'. it may only contain strings", command, t.String(), t.Type())
+		}
+	}
+
+	ret := liveUpdateRunStep{
+		command:  command,
+		triggers: triggerStrings,
+		position: thread.TopFrame().Position(),
+	}
+	s.unconsumedLiveUpdateSteps = append(s.unconsumedLiveUpdateSteps, ret)
+	return ret, nil
+}
+
+func (s *tiltfileState) liveUpdateRestartContainer(thread *starlark.Thread, fn *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	if err := starlark.UnpackArgs(fn.Name(), args, kwargs); err != nil {
+		return nil, err
+	}
+
+	ret := liveUpdateRestartContainerStep{
+		position: thread.TopFrame().Position(),
+	}
+	s.unconsumedLiveUpdateSteps = append(s.unconsumedLiveUpdateSteps, ret)
+	return ret, nil
+}
+
+func liveUpdateStepToModel(l liveUpdateStep, workDir string) (model.LiveUpdateStep, error) {
+	switch x := l.(type) {
+	case liveUpdateWorkDirStep:
+		return model.LiveUpdateWorkDirStep(x.value), nil
+	case liveUpdateSyncStep:
+		remotePath := x.remotePath
+		if !filepath.IsAbs(remotePath) {
+			if workDir == "" {
+				if !filepath.IsAbs(x.remotePath) {
+					return model.LiveUpdateSyncStep{}, fmt.Errorf("step '%s' must either follow a work_dir or have an absolute remote_path", x.String())
```

Comment:
again, we could also decide that if there's no workdir, we just treat workdir as "/" -- that seems more intuitive to me